File: java/core/src/main/java/network/aika/activations/InhibitoryActivation.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.activations;

import network.aika.Document;
import network.aika.bindingsignal.BSType;
import network.aika.bindingsignal.BindingSignal;
import network.aika.neurons.Neuron;
import network.aika.typedefs.ActivationDefinition;
import network.aika.typedefs.SynapseDefinition;

import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;
import java.util.stream.Stream;

/**
 *
 * @author Lukas Molzberger
 */
public class InhibitoryActivation extends Activation {

    protected NavigableMap<Integer, Link> inputLinks = new TreeMap<>();
    protected NavigableMap<Integer, Link> outputLinks = new TreeMap<>();

    public InhibitoryActivation(
            ActivationDefinition t,
            Activation parent,
            Integer id,
            Neuron n,
            Document doc,
            Map<BSType, BindingSignal> bindingSignals
    ) {
        super(t, parent, id, n, doc, bindingSignals);
    }

    @Override
    public void addInputLink(Link l) {
        int bsId = getInputKey(l);
        assert inputLinks.get(bsId) == null;
        inputLinks.put(bsId, l);
    }

    public Link getInputLink(int bsId) {
        return inputLinks.get(bsId);
    }

    public int getInputKey(Link l) {
        BSType wildcard = ((ActivationDefinition) type).getWildcard();
        BSType inputBSType = ((SynapseDefinition)l.getSynapse().getType()).mapTransitionBackward(wildcard);
        BindingSignal inputBS = l.getInput().getBindingSignal(inputBSType);
        return inputBS.getTokenId();
    }

    @Override
    public void addOutputLink(Link l) {
        int bsId = getOutputKey(l);
        assert outputLinks.get(bsId) == null;
        outputLinks.put(bsId, l);
    }

    public Link getOutputLink(int bsId) {
        return outputLinks.get(bsId);
    }

    public int getOutputKey(Link l) {
        BSType wildcard = ((ActivationDefinition) type).getWildcard();
        BSType outputBSType = ((SynapseDefinition)l.getSynapse().getType()).mapTransitionForward(wildcard);
        BindingSignal outputBS = l.getOutput().getBindingSignal(outputBSType);
        return outputBS.getTokenId();
    }

    @Override
    public void linkIncoming(Activation excludedInputAct) {

    }

    @Override
    public Stream<Link> getInputLinks() {
        return inputLinks.values()
                .stream();
    }

    @Override
    public Stream<Link> getOutputLinks() {
        return outputLinks.values()
                .stream();
    }

    @Override
    public Link getCorrespondingInputLink(Link l) {
        int bsId = getOutputKey(l);
        return getInputLink(bsId);
    }

    @Override
    public Link getCorrespondingOutputLink(Link l) {
        int bsId = getInputKey(l);
        return getOutputLink(bsId);
    }
}

File: src/network/inhibitory_activation.cpp
#include "network/inhibitory_activation.h"
#include "network/activation_definition.h"
#include "network/synapse_definition.h"
#include "network/binding_signal.h"
#include "fields/relation.h"
#include "fields/rel_obj_iterator.h"

InhibitoryActivation::InhibitoryActivation(ActivationDefinition* t, Activation* parent, int id, Neuron* n, Document* doc, std::map<BSType*, BindingSignal*> bindingSignals)
    : Activation(t, parent, id, n, doc, bindingSignals) {}
    
RelatedObjectIterable* InhibitoryActivation::followManyRelation(Relation* rel) const {
    if (rel->getRelationName() == "INPUT") {
        // Convert inputLinks to a vector of Obj*
        std::vector<Obj*> objs;
        for (const auto& pair : inputLinks) {
            objs.push_back(static_cast<Obj*>(pair.second));
        }
        return new VectorObjectIterable(objs);
    } else if (rel->getRelationName() == "OUTPUT") {
        // For InhibitoryActivation, we override both INPUT and OUTPUT handling
        std::vector<Obj*> objs;
        for (const auto& pair : outputLinks) {
            objs.push_back(static_cast<Obj*>(pair.second));
        }
        return new VectorObjectIterable(objs);
    } else {
        // Use base class implementation for other relations
        return Activation::followManyRelation(rel);
    }
}

void InhibitoryActivation::addInputLink(Link* l) {
    int bsId = getInputKey(l);
    assert(inputLinks.find(bsId) == inputLinks.end());
    inputLinks[bsId] = l;
}

Link* InhibitoryActivation::getInputLink(int bsId) const {
    auto it = inputLinks.find(bsId);
    if (it != inputLinks.end()) {
        return it->second;
    }
    return nullptr;
}

int InhibitoryActivation::getInputKey(Link* l) const {
    BSType* wildcard = static_cast<ActivationDefinition*>(type)->getWildcard();
    BSType* inputBSType = static_cast<SynapseDefinition*>(l->getSynapse()->getType())->mapTransitionBackward(wildcard);
    BindingSignal* inputBS = l->getInput()->getBindingSignal(inputBSType);
    return inputBS->getTokenId();
}

void InhibitoryActivation::addOutputLink(Link* l) {
    int bsId = getOutputKey(l);
    assert(outputLinks.find(bsId) == outputLinks.end());
    outputLinks[bsId] = l;
}

Link* InhibitoryActivation::getOutputLink(int bsId) const {
    auto it = outputLinks.find(bsId);
    if (it != outputLinks.end()) {
        return it->second;
    }
    return nullptr;
}

int InhibitoryActivation::getOutputKey(Link* l) const {
    BSType* wildcard = static_cast<ActivationDefinition*>(type)->getWildcard();
    BSType* outputBSType = static_cast<SynapseDefinition*>(l->getSynapse()->getType())->mapTransitionForward(wildcard);
    BindingSignal* outputBS = l->getOutput()->getBindingSignal(outputBSType);
    return outputBS->getTokenId();
}

void InhibitoryActivation::linkIncoming(Activation* excludedInputAct) {}

std::vector<Link*> InhibitoryActivation::getInputLinks() const {
    std::vector<Link*> result;
    for (const auto& pair : inputLinks) {
        result.push_back(pair.second);
    }
    return result;
}

std::vector<Link*> InhibitoryActivation::getOutputLinks() const {
    std::vector<Link*> result;
    for (const auto& pair : outputLinks) {
        result.push_back(pair.second);
    }
    return result;
}

Link* InhibitoryActivation::getCorrespondingInputLink(const Link* l) const {
    int bsId = getOutputKey(const_cast<Link*>(l));
    return getInputLink(bsId);
}

Link* InhibitoryActivation::getCorrespondingOutputLink(const Link* l) const {
    int bsId = getInputKey(const_cast<Link*>(l));
    return getOutputLink(bsId);
} 